### PR TITLE
Recurrent Batch Normalization

### PR DESCRIPTION
**Pull request by xingdi-eric-yuan**
Add batch normalization into recurrent networks.
Continue on @udibr 's work in [PR #2183](https://github.com/fchollet/keras/pull/2183)
Applied BN to SimpleRNN/GRU/LSTM

### CODE REVIEW

**Comment by xingdi-eric-yuan** [comment]
I simply tested the code using a TITAN X (CNMeM enabled). 
The test script I used is [lstm_benchmark](https://github.com/fchollet/keras/blob/master/examples/lstm_benchmark.py) (slightly modified, for testing all 3 types of RNN, and with/without BN)
Results are as follow:
SimpleRNN:
simplernn
GRU:
gru
LSTM:
lstm

**Comment by xingdi-eric-yuan** [comment]
@cooijmanstim could you take time check this?

**Comment by xingdi-eric-yuan** [comment]
well apparently it fails when using TF backend, which I'm not very familiar with. Any TF guy can help check it? especially [HERE](https://github.com/xingdi-eric-yuan/keras/blob/master/keras/layers/recurrent.py#L217)

**Comment by xingdi-eric-yuan** [comment]
I have changed the bn function to :

```
    def bn(self, x, _gamma, _beta):
        x_shape = K.shape(x)
        _gamma = K.expand_dims(_gamma, dim=0)
        _beta = K.expand_dims(_beta, dim=0)
        x = K.reshape(x, [x_shape[0], K.shape(_gamma)[1], -1])
        x_var = x - K.mean(x, axis=2, keepdims=True)
        x_std = K.sqrt(K.mean(K.square(x_var), axis=2, keepdims=True) + self.epsilon)
        res = x_var / x_std * _gamma + _beta
        res = K.reshape(res, x_shape)
        return res
```
but still has error, seems like the reshape part doesn't work in TF. Any ideas?

**Comment by xingdi-eric-yuan** [comment]
Also tested on cpu.
SimpleRNN on CPU:
simplernn_cpu
GRU on CPU:
gru_cpu
LSTM on CPU:
lstm_cpu

File: keras/layers/recurrent.py [hunk 1]
```diff
@@ -151,17 +153,28 @@ class Recurrent(Layer):
...
         self.go_backwards = go_backwards
         self.stateful = stateful
         self.unroll = unroll
         self.consume_less = consume_less
+        self.batch_norm = batch_norm
+        self.batch_norm_gamma = batch_norm_gamma
+        if self.batch_norm:
+            self.beta_init = initializations.get('zero')
+            self.epsilon = 1e-6
+            self.uses_learning_phase = True
```
Comment (fchollet) [inline keras/layers/recurrent.py]: You are setting this to True but don't seem to be using the learning phase in the code.

File: keras/layers/recurrent.py [hunk 2]
```diff
@@ -198,6 +214,16 @@ def get_initial_states(self, x):
         initial_states = [initial_state for _ in range(len(self.states))]
         return initial_states
 
+    def bn(self, x, _gamma, _beta, matrix_amount=1):
+        _gamma = K.expand_dims(_gamma, dim=0)
+        _beta = K.expand_dims(_beta, dim=0)
+        x = K.reshape(x, [-1, matrix_amount, self.output_dim])
+        x_var = x - K.mean(x, axis=2, keepdims=True)
+        x_std = K.sqrt(K.mean(K.square(x_var), axis=2, keepdims=True) + self.epsilon)
+        res = x_var / x_std * _gamma + _beta
```
Comment (fchollet) [inline keras/layers/recurrent.py]: This is per-batch normalization, usually used during training. You are not computing running averages of std and mean for normalization during testing. Why?

File: keras/layers/recurrent.py [hunk 3]
```diff
@@ -107,6 +107,9 @@ class Recurrent(Layer):
             in your model, you would need to specify the input length
             at the level of the first layer
             (e.g. via the `input_shape` argument)
+        batch_norm: Boolean (default False). If True, use batch Normalization,
```
Comment (fchollet) [inline keras/layers/recurrent.py]: Please don't capitalize "normalization"

File: keras/layers/recurrent.py [hunk 4]
```diff
@@ -151,17 +153,28 @@ class Recurrent(Layer):
...
         following the notes on statefulness RNNs.
+
+    # References
+        - [Recurrent Batch Normalization](https://arxiv.org/abs/1603.09025)
     '''
     def __init__(self, weights=None,
                  return_sequences=False, go_backwards=False, stateful=False,
                  unroll=False, consume_less='cpu',
-                 input_dim=None, input_length=None, **kwargs):
+                 input_dim=None, input_length=None, batch_norm=False, batch_norm_gamma=0.1, **kwargs):
```
Comment (fchollet) [inline keras/layers/recurrent.py]: Please put the new parameters on a new line

**Comment by fchollet** [comment]
Not convinced about implementation correctness.

File: keras/layers/recurrent.py (hunk 2 above)
Comment (xingdi-eric-yuan) [inline keras/layers/recurrent.py]: thanks for pointing it out, I'll fix that tonight

**Comment by xingdi-eric-yuan** [comment]
@fchollet I got some problem updating the running mean&std
[Here](https://github.com/xingdi-eric-yuan/keras/blob/master/keras/layers/recurrent.py#L249), what's an easy way to update stuff inside rnn scan?

**Comment by braingineer** [comment]
@xingdi-eric-yuan I actually have this problem (updating inside scan) all over the place and I haven't really gotten a great solution that jives with Keras mater branch.  Maybe one of the following ways will suit the problem. If you find another way, please let me know =). 

If you're using theano and it's a matrix, [T.set_subtensor is the preferred method](http://deeplearning.net/software/theano/tutorial/faq_tutorial.html).  you can see what I did [here to implement a stack rnn](https://github.com/braingineer/keras/blob/dev/keras/backend/theano_backend.py#L745).   In general, you should always [try to include all shared variables in the scan](http://deeplearning.net/software/theano/library/scan.html#using-shared-variables-the-strict-flag). 

if it's a scalar, another option is to include it in the loop  as a passed variable.. [here I did it for a clockwork rnn](https://github.com/braingineer/ikelos/blob/master/ikelos/layers/cwrnn.py#L102)

**Comment by udibr** [comment]
Hi @xingdi-eric-yuan I'm sorry but it looks to me like you started your work on the wrong branch. As I [said before](https://github.com/fchollet/keras/pull/2183#issuecomment-217528846) I have a working example of LSTM BN including the update step. Yes it is complex work but it is doable.
The code is [here](https://github.com/udibr/keras/blob/mymaster/keras/layers/recurrent.py#L583) 

If you look at the code you will see that in the [step function](https://github.com/udibr/keras/blob/mymaster/keras/layers/recurrent.py#L883) I carefully take the mean/std from previous step and unpack it and then after computing the updated mean and std I send it [back](https://github.com/udibr/keras/blob/mymaster/keras/layers/recurrent.py#L933)

In other words things that are being updated by an RNN step (or anything which uses a scan) needs to be passed as part of the states of the step. Different value on input and output. This was already in use in the original LSTM for the states if the cell  and the same should be applied to anything else (mean or std) which is updated in the step

**Comment by udibr** [comment]
As I also said before this is a very messy code and I dont like it. Also the graph you produced did not show any advantage for using BN in LSTM so perhaps you need to find a different test program

**Comment by braingineer** [comment]
(sorry for the side topic questions)
re state passing @udibr :
- so [this line](https://github.com/udibr/keras/blob/mymaster/keras/layers/recurrent.py#L769), the `replace_row` is your update, right?  And it's just a set_subtensor on theano side?  
-  concerning the TF side, how would your solution look if you were updating a matrix slice rather than row slice?   (and is this something that could be PRed back into keras master?)

**Comment by xingdi-eric-yuan** [comment]
Thanks guys, @braingineer @udibr 
@udibr , I was trying to find out is there any better way to do updates, and I guess I'll have to use similar way you used. however because of the consume_less='gpu' mode, maybe it's better to do the calculations with gamma&beta in a batched style, instead of separately, and that's exactly my concern. I'll continue think about this.
@braingineer yes, his [replace_row](https://github.com/udibr/keras/blob/mymaster/keras/backend/theano_backend.py#L406) is `T.set_subtensor`

**Comment by xingdi-eric-yuan** [comment]
Also @udibr , 
> Also the graph you produced did not show any advantage for using BN in LSTM so perhaps you need to find a different test program

After investigation, I found there was a stupid bug in my test code, so I removed the graphs, I'll re-run it after I finish this. Sorry about that.

**Comment by udibr** [comment]
@braingineer I did add the replace_row and I see I have an implementation in the [tf backend](https://github.com/udibr/keras/blob/mymaster/keras/backend/tensorflow_backend.py#L576)

**Comment by udibr** [comment]
@braingineer also you can always convert a matrix to a row with reshape, do your replace and then convert back to matrix with another reshape

**Comment by braingineer** [comment]
@udibr ya, I had seen both of those, but I didn't seen an immediately obvious way of expanding it to matrices, as it is to do in theano.  

re: reshaping. hmm. maybe. seems hacky, but then again, updating tensors at all mid-computation is fairly hacky :\.

**Comment by udibr** [comment]
@braingineer you can do the reshape you your code using K.reshape so the hacking remains in your code

**Comment by braingineer** [comment]
> so the hacking remains in your code

I meant hacky in terms of computation.  doing a reshape no matter what results in an extra computation upon, which, modifying a tensor is already expensive.  though, maybe it doesn't matter and I'm bikeshedding.

**Comment by udibr** [comment]
I think reshape is super fast because it only changes the stride of the indexing but I may be wrong because even the shape is a theano variable and who knows what is going on there also theano may copy after all if memory is fragmented...

**Comment by xingdi-eric-yuan** [comment]
@udibr sorry I'm kind of confuse about your code, 
In [bn()](https://github.com/udibr/keras/blob/mymaster/keras/layers/recurrent.py#L874) you update `out_step_running_mean ` and `out_step_running_std `, and in [add_bn_to_states()](https://github.com/udibr/keras/blob/mymaster/keras/layers/recurrent.py#L764) you add `running_mean ` and `running_std ` into states, how do you update `out_step_running_xxx` into `running_xxx` ?

**Comment by udibr** [comment]
good catch! looks like a bug to me (it happens) very interesting to see how it will change effects after you fix it.

 I suggest adding running_mean and running_std as parameters to add_bn_to_states and in all calls passing the values self.running_mean and self.running_std except for the one time in which add_bn_to_states is called at end of step in which case out_step_running_xxx should be used.

**Comment by udibr** [comment]
@xingdi-eric-yuan I've did the bug fix for the problem you found and I moved my code back to the [lstmbn branch of my code](https://github.com/udibr/keras/blob/lstmbn/keras/layers/recurrent.py#L863)
I also put all my changes in a separate class `LSTMBN` so there will be no conflicts with keras master LSTM which is all the time changing.

**Comment by xingdi-eric-yuan** [comment]
@udibr cool, I'll look into the `LSTMBN` class later, busy week 👍

**Comment by dgboy2000** [comment]
@fchollet It looks like BN for LSTM is unfinished -- are you still interested in having this in Keras? I'd like to use it personally, and passing your standards for merging would help ensure correctness.
